### DISCUSSION TITLE
Changing the default layer names

### DISCUSSION THREAD

**Discussion by Qvadi** [iqm-finland/KQCircuits#75] [Q&A] [answered]
Hi KQC team,
I was trying to look into changing existing layer names and I think this can be done in the layer_config --> default_layer_config.py file. Changing here will apply the new layer names globally or would it require some additional work in making it work?

**Comment by qpavsmi** [comment, marked as answer]
Hi Qvadi,

Everything related to layer configuration is indeed under the `klayout_package/python/kqcircuits/layer_config` folder. Let me introduce the files that are there:

- `default_layer_props.lyp` a file used by KLayout to load the layer properties, to display layer with correct colors, transparency, dithering etc. This can be loaded in KLayout from top toolbar using `File > Load Layer Properties`. This file is messy xml style file that shouldn't really be modified by hand, unless all other options fail.
- `default_layer_config.py` a file to make the KQC code work. In our workflow we group a set of layers and duplicate them over multiple faces of our chip stack. Take for example `1t1` face. The first number tells us which order of the chip the face is on from bottom up, the letter tells us 't' for top face and 'b' for bottom face, then the last number allows to define multiple "faces" on one side. So this file defines which layers we can work with, everything outside of these layers we don't work on (some exceptions may apply when exporting GDS files for simulations)
- `example_layer_config.py` an example of what an alternative to `default_layer_config.py` could look like. This seems to be a bit out-of-date but hopefully should give an idea on how your own custom configuration could look like.

The configuration code that KQC uses can be configured from the `klayout_package/python/kqcircuits/defaults.py`, you can change `layer_config_path` to point to some other file than `default_layer_config.py`. So a good way to start is to copy the contents of `default_layer_config.py` into your own custom file `my_layer_config.py` and change `layer_config_path` to point to it.

Then we have a tool to generate the `*.lyp` file based on the `my_layer_config.py` file you made, it is `util/check_layer_props.py`. Running this in terminal and giving a path to your `my_layer_config.py` file will look for mismatches between `*.lyp` and `my_layer_config.py` files and report them. You can also run the script in "write" (`-w`) mode, in which case the script becomes interactive, prompting you to make changes in the `*.lyp` file if a mismatch is detected.

So a recommended way after making your own `*_layer_config.py` file is to:
1. Copy contents of `default_layer_props.lyp` into your own separate file `my_layer_props.lyp`
2. Run in terminal `python util/check_layer_props.py klayout_package/python/kqcircuits/layer_config/my_layer_props.lyp -w`
3. Double check the mismatches it detects. If you trust the tool, just say yes to everything

Then in KLayout you can open `my_layer_props.lyp` with `File > Load Layer Properties` and see in the right-most `Layers` pane if the layers are how you expect them.

For more reading check out https://iqm-finland.github.io/KQCircuits/user_guide/python_workflow.html#kqcircuit-s-layers chapters `KQCircuit’s Layers` and `Faces`.

# Alternative way, modifying my_layer_props.lyp directly

You can also modify `my_layer_props.lyp` more directly using KLayout but without opening the file in text editor.
 
![image](https://github.com/iqm-finland/KQCircuits/assets/109152229/eeafdcbf-4911-411c-98bd-57949babb3aa)

Right-clicking on the `Layers` pane there are sets of operations you can do to add, remove, modify, group etc layers currently loaded into KLayout. Though for adding new layers I would recommend to do it from top toolbar `Edit > Layer > New Layer`. You can use these utilities within KLayout to configure the layers how you like it, then `File > Save Layer Properties` to create your own `*.lyp` file that can be loaded after KLayout restart.
